engine/sim/sc_expressions.cpp

```diff
@@ -16,6 +16,145 @@ namespace
 {  // ANONYMOUS ====================================================
 
 constexpr bool EXPRESSION_DEBUG = false;
+
+struct lexer_t
+{
+  struct token_t {
+    token_e type = TOK_UNKNOWN;
+    util::string_view str;
+  };
+
+  util::string_view input;
+  size_t current_len = 0;
+
+  explicit lexer_t( util::string_view s )
+    : input( s )
+  {}
+
+  char peek() const
+  {
+    return current_len < input.size() ? input[ current_len ] : '\0';
+  }
+
+  bool match( char c )
+  {
+    if ( starts_with( input.substr( current_len ), c ) ) {
+      current_len++;
+      return true;
+    }
+    return false;
+  }
+
+  bool match( util::string_view str, int consume_mod = 0 )
+  {
+    if ( util::str_prefix_ci( input.substr( current_len ), str ) ) {
+      current_len += ( str.size() + consume_mod );
+      return true;
+    }
+    return false;
+  }
+
+  token_t yield_token( token_e type )
+  {
+    token_t token;
+    token.type = type;
+    token.str = input.substr( 0, current_len );
+    input.remove_prefix( current_len );
+    return token;
+  }
+
+  token_t next()
+  {
+    if ( input.empty() )
+      return yield_token( TOK_EOF );
+
+    const char ch = input.front();
+    current_len = 1;
+
+    // numbers
+    if ( is_digit( ch ) ) {
+      while ( is_digit( peek() ) )
+        current_len++;
+      if ( match( '.' ) ) {
+        while ( is_digit( peek() ) )
+          current_len++;
+      }
+      return yield_token( TOK_NUM );
+    }
+
+    // special case floor & ceil "pseudo-functions"
+    if ( lowercase( ch ) == 'f' && match( "loor(", -1 ) )
+      return yield_token( TOK_FLOOR );
+    if ( lowercase( ch ) == 'c' && match( "eil(", -1 ) )
+      return yield_token( TOK_CEIL );
+
+    // "strings" (identifiers, really)
+    if ( is_alpha( ch ) ) {
+      for ( char c = peek(); is_alpha( c ) || is_digit( c ) || c == '_' || c == '.'; c = peek() )
+        current_len++;
+      return yield_token( TOK_STR );
+    }
+
+    switch ( ch )
+    {
+      case '(': return yield_token( TOK_LPAR );
+      case ')': return yield_token( TOK_RPAR );
+      case '+': return yield_token( TOK_PLUS );
+      case '-': return yield_token( TOK_MINUS );
+      case '*': return yield_token( TOK_MULT );
+      case '%':
+      {
+        if ( match( '%' ) )
```

Comment:
So %% is mod, and a single % is division? Might be worth a comment since this is quite a bit unintuitive at first glance.

Reply:
Yeah, been like that since d68425e69c0c032053b7cafe43cfa8727596c78c
Will add (I think it's kinda self documenting, but I see how it can be non-obvious when looking at it for the first time).

Reply:
Added a comment for why we use `%`/`%%` for div/mod in b941c44